geo/src/algorithm/bool_ops/op.rs

```diff
@@ -0,0 +1,281 @@
+            debug_assert!(botmost_start_segment.at_left);
+
+            trace!(
+                "Bottom most start-edge: {botmost:?}",
+                botmost = botmost_start_segment.line,
+            );
+
+            let prev = iter.prev_active(&botmost_start_segment);
+            trace!(
+                "prev-active(bot-most): {prev:?}",
+                prev = prev.map(|(_, p)| p.geom)
+            );
+
+            let mut region = prev
+                .as_ref()
+                .map(|(g, c)| c.get_region(*g))
+                .unwrap_or_else(|| Region::infinity(self.ty));
+            trace!("bot region: {region:?}");
+
+            while idx < iter.intersections().len() {
+                let mut c = &iter.intersections()[idx];
+                let mut jdx = idx;
+                loop {
+                    region.cross(c.cross.is_first);
+                    let has_overlap = (idx + 1) < iter.intersections().len()
+                        && compare_crossings(c, &iter.intersections()[idx + 1]) == Ordering::Equal;
+                    if !has_overlap {
+                        break;
+                    }
+                    idx += 1;
+                    c = &iter.intersections()[idx];
+                }
+                trace!(
+                    "set_region: {geom:?} / {geom2:?} => {region:?} ({c} counts)",
+                    geom2 = c.cross.geom,
+                    geom = iter.intersections_mut()[jdx].line,
+                    c = idx - jdx + 1,
+                );
+                while jdx <= idx {
+                    let gpiece = iter.intersections()[jdx].line;
+                    iter.intersections()[jdx].cross.set_region(region, gpiece);
+                    jdx += 1;
+                }
+                idx += 1;
+            }
+        }
+
+        rings.finish()
+    }
+}
+
+#[derive(Clone, Copy)]
+struct Region {
+    is_first: bool,
+    is_second: bool,
+}
+impl Debug for Region {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "[{f}{s}]",
+            f = if self.is_first { "A" } else { "" },
+            s = if self.is_second { "B" } else { "" },
+        )
+    }
+}
+
+impl Region {
+    fn infinity(ty: OpType) -> Self {
+        Region {
+            is_first: false,
+            is_second: matches!(ty, OpType::Difference),
+        }
+    }
+    fn cross(&mut self, first: bool) {
+        if first {
+            self.is_first = !self.is_first;
+        } else {
+            self.is_second = !self.is_second;
+        }
+    }
+    fn is_ty(&self, ty: OpType) -> bool {
+        match ty {
+            OpType::Intersection | OpType::Difference => self.is_first && self.is_second,
+            OpType::Union => self.is_first || self.is_second,
+            OpType::Xor => self.is_first ^ self.is_second,
+        }
+    }
+}
+
+#[derive(Clone)]
+struct Edge<T: Float> {
+    geom: LineOrPoint<T>,
+    is_first: bool,
```

Comment:
I confess that I'm confused about the distinction between  edge._region.is_first vs edge.is_first - can you elucidate?

Reply:
An edge belongs to the first-input geometry or the second, so we use a flag.  A region is space between edges, and can belong to either/both/none of the geometries: 4 combinations -> two flags within Region.